src/main/java/com/google/sps/servlets/MentorEvidenceServlet.java

```diff
@@ -37,72 +37,84 @@
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    int userId = Utility.getUserId();
+    int mentorId = Utility.getUserId();
     
     // Get variable from HTML form.
     String paragraph = request.getParameter("paragraph");
 
+    // Delete existing notifications and approvers to re-create them, in case mentor is re-applying.
+    deleteNotifications(mentorId);
+    deleteApprovers(mentorId);
+
     // Update mentor evidence and add approvers in database.
-    updateMentorEvidence(userId, paragraph);
-    addApprovers(userId); //TODO(oumontiel): Only call when there are no approvers.
+    updateMentorEvidence(mentorId, paragraph);
+    addApprovers(mentorId); //TODO(oumontiel): Only call when there are no approvers.
 
     // Call NotificationServlet to notify approvers.
     response.setContentType("text/plain");
     try {
       request.getRequestDispatcher("/notification?type=requestApproval&modifiedElementId="
-          + userId).include(request, response);
+          + mentorId).include(request, response);
     } catch (ServletException exception) {
       Logger logger = Logger.getLogger(MentorApprovalServlet.class.getName());
       logger.log(Level.SEVERE, exception.getMessage(), exception);
     }
     response.sendRedirect("/index.html");
   }
 
+  /**
+   * Deletes existing notifications related to current mentor's approval from database.
+   */
+  private void deleteNotifications(int mentorId) {
+    // Set up query to delete all relations of users to notifications of this mentor approval.
+    String query = "DELETE FROM UserNotification "
+        + "WHERE notification_id IN "
+        + "(SELECT id FROM Notification "
+        + "WHERE url = '/approval.html?id=" + mentorId + "')";
+    Utility.executeQuery(query);
+
+    // Set up query to delete all notifications related to mentor.
+    query = "DELETE FROM Notification "
+        + "WHERE url = '/approval.html?id=" + mentorId + "'";
+    Utility.executeQuery(query);
+  }
```

Comment:
maybe you could make the function more generic by passing the url as a parameter and moving it to Utility, since Andres could also use it for deleting post notifications, but that's up to you

Reply:
But he would require a different function, because this includes a specific table that has foreign keys to the Notification table, which other types of notifications don't have, so it wouldn't work (unless I applied `ON DELETE CASCADE` which I couldn't do). Just in case he ever makes a Utility method to delete a generic notification, I change its name to `deleteApprovalNotifications`.